src/storage/src/source/postgres/replication.rs

```diff
@@ -0,0 +1,556 @@
+        loop {
+            let send_feedback = match select(stream.as_mut().next(), uppers.as_mut().next()).await {
+                future::Either::Left((next_message, _)) => match next_message.transpose()? {
+                    Some(ReplicationMessage::XLogData(data)) => {
+                        yield data.into_data();
+                        last_feedback.elapsed() > FEEDBACK_INTERVAL
+                    }
+                    Some(ReplicationMessage::PrimaryKeepAlive(keepalive)) => keepalive.reply() == 1,
+                    Some(_) => Err(anyhow!("unexpected replication message"))?,
+                    None => return,
+                },
+                future::Either::Right((upper, _)) => match upper.and_then(|u| u.into_option()) {
+                    Some(lsn) => {
+                        last_committed_upper = std::cmp::max(last_committed_upper, lsn);
+                        true
+                    }
+                    None => false,
+                },
+            };
+            if send_feedback {
+                let ts: i64 = PG_EPOCH.elapsed().unwrap().as_micros().try_into().unwrap();
+                // For the same reason as above, we must subtract one here too.
+                let lsn = PgLsn::from(last_committed_upper.offset.saturating_sub(1));
+                stream
+                    .as_mut()
+                    .standby_status_update(lsn, lsn, lsn, ts, 0)
+                    .await?;
+                last_feedback = Instant::now();
+            }
+        }
+    })
+}
+
+/// Extracts a single transaction from the replication stream delimited by a BEGIN and COMMIT message.
+fn extract_transaction<'a>(
+    stream: impl AsyncStream<Item = Result<LogicalReplicationMessage, anyhow::Error>> + 'a,
+    table_info: &'a BTreeMap<u32, (usize, PostgresTableDesc, Vec<MirScalarExpr>)>,
+    connection_config: &'a mz_postgres_util::Config,
+    metrics: &'a PgSourceMetrics,
+    publication: &'a str,
+    errored_tables: &'a mut HashSet<u32>,
+) -> impl AsyncStream<
+    Item = Result<(u32, Result<Vec<Option<Bytes>>, DefiniteError>, Diff), anyhow::Error>,
+> + 'a {
+    use LogicalReplicationMessage::*;
+    async_stream::try_stream!({
+        let mut stream = pin!(stream);
+        let commit_lsn = match stream.try_next().await? {
+            Some(Begin(body)) => body.final_lsn(),
+            Some(_) => Err(anyhow!("transaction begun without BEGIN"))?,
+            None => Err(anyhow!("stream ended prematurely"))?,
+        };
+        metrics.transactions.inc();
+        metrics.lsn.set(commit_lsn.into());
+        while let Some(event) = stream.try_next().await? {
+            metrics.total.inc();
+            match event {
+                Insert(body) if errored_tables.contains(&body.rel_id()) => continue,
+                Update(body) if errored_tables.contains(&body.rel_id()) => continue,
+                Delete(body) if errored_tables.contains(&body.rel_id()) => continue,
```

Comment:
Should also have one of these checks for `Relation`

Reply:
Good catch! fixed